From: Jeremy Lainé <jeremy.laine@m4x.org>
Subject: relax neon version check
Bug: https://issues.asterisk.org/jira/browse/ASTERISK-24325

Relax the neon version check to also accept version 0.30.x

diff --git a/res/res_calendar_ews.c b/res/res_calendar_ews.c
index 0f8b0c2..a617790 100644
--- a/res/res_calendar_ews.c
+++ b/res/res_calendar_ews.c
@@ -907,7 +907,7 @@ static int load_module(void)
 {
 	/* Actualy, 0.29.1 is required (because of NTLM authentication), but this
 	 * function does not support matching patch version. */
-	if (ne_version_match(0, 29)) {
+	if (ne_version_match(0, 29) && ne_version_match(0, 30)) {
 		ast_log(LOG_ERROR, "Exchange Web Service calendar module require neon >= 0.29.1, but %s is installed.\n", ne_version_string());
 		return AST_MODULE_LOAD_DECLINE;
 	}
